pkl-core/src/main/java/org/pkl/core/runtime/VmMap.java

```diff
@@ -219,12 +225,39 @@ public void force(boolean allowUndefinedValues) {
     }
   }
 
+  @TruffleBoundary
+  public VmMapping toMapping() {
+    var builder = new VmObjectBuilder(keyOrder.size());
+    for (var key : keyOrder) {
+      var value = map.get(key);
+      assert value != null;
+      builder.addEntry(key, value);
+    }
+    return builder.toMapping();
+  }
+
+  @TruffleBoundary
+  public VmDynamic toDynamic() {
+    var builder = new VmObjectBuilder(keyOrder.size());
+    for (var key : keyOrder) {
+      var value = map.get(key);
+      assert value != null;
+      if (key instanceof String) {
+        builder.addProperty(Identifier.get((String) key), value);
+      } else {
+        builder.addEntry(key, value);
+      }
+    }
+    return builder.toDynamic();
+  }
+
```

Comment:
This is a replacement for code that was visible to Truffle inlining, so it shouldn't be then moved behind a `TruffleBoundary`.

This works:

```suggestion
  public VmMapping toMapping() {
    var builder = new VmObjectBuilder(getLength());
    for (var entry : this) {
      builder.addEntry(VmUtils.getKey(entry), VmUtils.getValue(entry));
    }
    return builder.toMapping();
  }

  public VmDynamic toDynamic() {
    var builder = new VmObjectBuilder(getLength());
    for (var entry : this) {
      var key = VmUtils.getKey(entry);
      var value = VmUtils.getValue(entry);
      if (key instanceof String) {
        builder.addProperty(Identifier.get((String) key), value);
      } else {
        builder.addEntry(key, value);
      }
    }
    return builder.toDynamic();
  }

```